### PR TITLE
#165020135 Add search filter for articles

### PR DESCRIPTION
**Pull request by coolbeatz71**
#### What does this PR do?
- Add search filter for articles

#### Description of Task to be completed?
- Add controller to support filter by author, title, tag and keyword
- Add pagination support for search filter with offset and limit
- Add the route to support search filter
- Add validation middleware for each attributes (author, title, tag and keyword)
- Document the api with swagger
- Allow user to search article with `many tags` separated by commas or space
#### How should this be manually tested?

- npm test
- with Postman

#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
#165020135
#### Screenshots (if appropriate)
~ Request screenshot on postman

tag
title
author
keyword

~ Response screenshot on postman

res_tag
res_title
res_author
res_keyword

#### Questions:
- Should the endpoint also support combination of filters? (need clarification)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -14,9 +14,9 @@
 		"build": "babel app.js --out-dir build",
 		"dev": "NODE_ENV=development nodemon  --require @babel/register app.js",
 		"coverage": "nyc report --reporter=text-lcov | coveralls",
-		"dbInit-dev": "NODE_ENV=development sequelize db:create || NODE_ENV=development sequelize db:migrate",
-		"dbInit-test": "NODE_ENV=test npx sequelize db:create || NODE_ENV=test npx sequelize db:migrate",
-		"dbInit": "npm run dbInit-dev || npm run dbInit-test"
+		"dbInit-dev": "NODE_ENV=development sequelize db:create && NODE_ENV=development sequelize db:migrate",
```
Comment (dusmel) [inline package.json]: Hey @coolbeatz71 why did you change these lines, we chose `||`so that when the db is already created the migration doesn't fail. when using `&&` the creation of db must be successful for the migration to run

File: server/controllers/articles.js [hunk 2]
```diff
@@ -317,5 +320,206 @@ class Articles {
...
+    } = req.query;
+
+    const pagination = { offset, limit };
+
+    try {
+      if (author) {
+        Articles.filterByAuthor(res, author, pagination);
+      } else if (title) {
+        Articles.filterByTitle(res, title, pagination);
+      } else if (keyword) {
```
Comment (dusmel) [inline server/controllers/articles.js:457]: I think keyword and tag is just one thing
image

File: server/controllers/articles.js [hunk 3]
```diff
@@ -317,5 +320,206 @@ class Articles {
...
+          },
+          {
+            model: Rating,
+            attributes: ['rate']
+          }
+        ],
+        order: [['createdAt', 'DESC']],
+      }]
+    });
+
```
Comment (dusmel) [inline server/controllers/articles.js:614]: What if we respond with an appropriate message when the result doesnt return anything instead of returning an empty array? can also apply to article title and author

File: server/test/articles.spec.js [hunk 4]
```diff
@@ -309,4 +317,143 @@ describe('Article endpoints', () => {
...
+          .get('/api/search?author=')
+          .end((error, res) => {
+            expect(res.status).to.be.equal(400);
+            expect(res.body).to.have.property('message');
+            expect(res.body.message).to.have.equals('author is not allowed to be empty');
+            done();
+          });
+      });
+    });
+
```
Comment (dusmel) [inline server/test/articles.spec.js:578]: I would suggest testing also if the author does not exist, or tag or article title.

File: package.json (hunk 1 above)
Comment (coolbeatz71) [reply package.json]: Yup @dusmel. I understand your concern. I changed it because, in case we need to prepare at the same time the test and dev DB, this oblige us to run the script twice. Because of this `||`. But I also noticed that using `&&` is preventing the script to finish in the case where the DB already exist. Leading us to an error. The solution for this issue is to use `&` which has the advantage of running both script simultaneously and when the DB already exist, it is not preventing the rest of the script to be executed.

### **See this screenshot**
Screen Shot 2019-05-07 at 17 35 04

File: server/controllers/articles.js (hunk 2 above)
Comment (coolbeatz71) [reply server/controllers/articles.js:457]: > I think keyword and tag is just one thing
> image

Hey @dusmel. Thank you for the feedback. When I have seen the term `keyword`, I personally understood that `it can be any word or sentence inside the article or which is related to the article.` So I am actually searching the keyword in the article title, body or article description. I should also add the case where we can search the keyword inside the tag.

File: server/controllers/articles.js (hunk 3 above)
Comment (coolbeatz71) [reply server/controllers/articles.js:614]: > What if we respond with an appropriate message when the result doesn't return anything instead of returning an empty array? can also apply to article title and author

@dusmel, I think the frontend should handle this case as we agreed on using `infinite scrolling`. The backend should not respond with a message. An empty array means `no data to display`. In brief, the frontend will display a message in case the article array is empty

File: server/test/articles.spec.js (hunk 4 above)
Comment (coolbeatz71) [reply server/test/articles.spec.js:578]: > I would suggest testing also if the author does not exist, or tag or article title.

@dusmel. If the user input an author who doesn't exist, at least we are searching author with a name which `looks like` it. In case no name looks like the author name. We are displaying an empty array. For the  `empty array` the solution is here https://github.com/andela/ah-kgl-wakanda-backend/pull/30#discussion_r281453906
